govcd/vm_affinity_rule.go

```diff
@@ -0,0 +1,397 @@
+	}
+	getById := func(id string, refresh bool) (interface{}, error) { return vdc.GetVmAffinityRuleById(id) }
+	entity, err := getEntityByNameOrId(getByName, getById, identifier, false)
+	if entity == nil {
+		return nil, err
+	}
+	return entity.(*VmAffinityRule), err
+}
+
+func validateAffinityRule(client *Client, affinityRuleDef *types.VmAffinityRule, checkVMs bool) (*types.VmAffinityRule, error) {
+	if affinityRuleDef == nil {
+		return nil, fmt.Errorf("empty definition given for a VM affinity rule")
+	}
+	if affinityRuleDef.Name == "" {
+		return nil, fmt.Errorf("no name given for a VM affinity rule")
+	}
+	if affinityRuleDef.Polarity == "" {
+		return nil, fmt.Errorf("no polarity given for a VM affinity rule")
+	}
+	if !correctPolarity(affinityRuleDef.Polarity) {
+		return nil, fmt.Errorf("illegal polarity given (%s) for a VM affinity rule", affinityRuleDef.Polarity)
+	}
+	//if len(affinityRuleDef.VmReferences) == 0 || len (affinityRuleDef.VmReferences[0].VMReference) <2 {
+	//	return nil, fmt.Errorf("at least 2 VMs should be given for a VM Affinity Rule")
+	//}
+	// Ensure the VMs in the list are different
+	var seenVms = make(map[string]bool)
+	var allVmMap = make(map[string]bool)
+	if checkVMs {
+		vmList, err := client.QueryVmList(types.VmQueryFilterOnlyDeployed)
+		if err != nil {
+			return nil, fmt.Errorf("error getting VM list : %s", err)
+		}
+		for _, vm := range vmList {
+			allVmMap[extractUuid(vm.HREF)] = true
+		}
+	}
+	for _, vmr := range affinityRuleDef.VmReferences {
+		if len(vmr.VMReference) == 0 {
+			continue
+		}
+		for _, vm := range vmr.VMReference {
+			if vm == nil {
+				continue
+			}
+			// The only mandatory field is the HREF
+			if vm.HREF == "" {
+				return nil, fmt.Errorf("empty VM HREF provided in VM list")
+			}
+			_, seen := seenVms[vm.HREF]
+			if seen {
+				return nil, fmt.Errorf("VM HREF %s used more than once", vm.HREF)
+			}
+			seenVms[vm.HREF] = true
+
+			if checkVMs {
+				// Checking that the VMs indicated exist.
+				// Without this check, if any of the VMs do not exist, we would get an ugly error that doesn't easily explain
+				//  the nature of the problem, such as
+				//   > "error instantiating a new VM affinity rule: API Error: 403: [ ... ]
+				//   > Either you need some or all of the following rights [ORG_VDC_VM_VM_AFFINITY_EDIT]
+				//   > to perform operations [VAPP_VM_EDIT_AFFINITY_RULE] for $OP_ID or the target entity is invalid"
+
+				_, vmInList := allVmMap[extractUuid(vm.HREF)]
+				if !vmInList {
+					return nil, fmt.Errorf("VM identified by '%s' not found ", vm.HREF)
+				}
+			}
+		}
+	}
+	if len(seenVms) < 2 {
+		return nil, fmt.Errorf("at least 2 VMs should be given for a VM Affinity Rule")
+	}
+	return affinityRuleDef, nil
+}
+
+func (vdc *Vdc) CreateVmAffinityRuleAsync(affinityRuleDef *types.VmAffinityRule) (Task, error) {
+
+	var err error
+	// We validate the input, without a strict check on the VMs
```

Comment:
what means strict check in this context?

Reply:
It means that, on first pass, we don't check that the VMs passed in the list are valid.
We do such check only when we get an error during creation.

1. creation
2. if no error -> end
3. if error, validate VMs
  3a. if validation error, it was a VM problem: return combined error + list of VM that are invalid
  3b. if not validation error, it was a different issue: return original error